### PR TITLE
Add support for enforcing CRL expiration using nextUpdate field

### PR DESCRIPTION
**Pull request by jasperpatterson**
This adds support for enforcing the CRL nextUpdate field (i.e. expiration). Currently there is no way for a consumer of this crate to enforce expiration on a CRL since the nextUpdate field is not even exposed.

I considered simply exposing the `next_update` field on the `CertRevocationList` for a consumer to use on their own, but opted for an expiration enforcement option which seems more convenient, and follows the pattern of other builder options that already exist on `RevocationOptionsBuilder`.

I found this idea was also mentioned in one of the [original PRs that added support for CRL](https://github.com/rustls/webpki/pull/66):
> There is no enforcement of the current time being greater than the CRL's NextUpdate. I found most implementations I surveyed chose not to treat this as a hard error so for now I've omitted any enforcement. Maybe this should be configurable?

To enforce expiration, you pass configuration to the `RevocationOptionsBuilder`:

```rust
RevocationOptionsBuilder::new(&crls)
	.unwrap()
	.with_expiration_policy(ExpirationPolicy::Enforce)
	.build();
```

The default behaviour, ignoring the nextUpdate field, remains unchanged.

### CODE REVIEW

**Review by djc — APPROVED** [review]
LGTM, thanks!

The policy of ignoring by default feels at odds a little with our usual default of failing closed... Do we have any context of why other implementations aren't enforcing this?

@jasperpatterson for the clippy lint, probably just suppress it for that function?

**Comment by jasperpatterson** [comment]
> The policy of ignoring by default feels at odds a little with our usual default of failing closed... Do we have any context of why other implementations aren't enforcing this?

Personally, I agree. I expected this to be enforced and only caught that it wasn't because I wrote a test covering it.

The closest explanation I could find was this comment (about OCSP, not CRL) in the Go library: https://github.com/golang/go/issues/45244#issuecomment-810538506

I also was unsure about changing the default here since it could be a breaking change.

**Comment by djc** [comment]
It might be good to look at what other implementations (notably BoringSSL and the Go std library) do for this case?

File: src/crl/mod.rs
Comment (ctz) [inline src/crl/mod.rs:148]: This should be below the signature check -- the nextUpdate field is inside TBS and should be treated as indeterminate until the signature is validated.

**Review by ctz — COMMENTED** [review]
I think this would benefit from some integration test cases to explore the actual behaviour we want:

1) a single CRL authoritative for the certificate, and the verification epoch is past nextUpdate.
2) two CRLs authoritative for the certificate, one past, one current. I think the code right now will select the first one, and ignore the other. That is not ideal in several ways: a) later updates to the CRL are ignored, b) it fixates on an expired CRL when a "better" one is available, c) the order of CRLs in our API is unspecified, I think.

I think maybe it would be a better idea to do the expiration inside `authoritative()` which would work correctly for (2), while falling to `status_policy` for rejecting the certificate.

**Comment by cpu** [comment]
> It might be good to look at what other implementations (notably BoringSSL and the Go std library) do for this case?

AIUI the Go standard library only exposes creating and verifying the signatures of CRLs, not processing them for revocation decisions so it's not a very helpful point of comparison.

I looked at https://github.com/grpc/grpc-go's "advancedtls" package as one data point back when I was starting this work and found that it did check the CRL next update to decide if a cached CRL was appropriate or not, but if it was expired and the on-disk copy wasn't updated, it would proceed using the expired CRL. I emailed the maintainers at the time to see if they considered this a bug/security issue and they replied that it was by design saying:

> Many implementations treat NextUpdate as an expiry, but we don't think this is always a good choice. If you choose to fail-open, an "old" CRL should be more secure than no CRL. And if you fail-closed, an "old" CRL would mean you reject everything, which is certainly secure but probably not desirable in many cases.

After that I compared with other implementations and found they acted similar and I replied:

> Reading through BoringSSL's CRL processing code I believe it doesn't hard fail on expired CRLs and instead uses the validity as one heuristic input for a score used to determine the appropriateness of consulting the CRL. It's tricky from my cursory scan to determine if expiration alone would ever tip the scales but seems closer to your interpretation.
> 
> The handling within s2n [is clearer](https://github.com/aws/s2n-tls/blob/1a184431c1c756d8476a1aba4e0ebc32ea502ec5/tls/s2n_crl.c#L239-L243) and appears to explicitly eat the X509_V_ERR_CRL_HAS_EXPIRED error, returning 1 from the X509_STORE_CTX_set_verify_cb.

**Comment by jasperpatterson** [comment]
> two CRLs authoritative for the certificate, one past, one current. I think the code right now will select the first one, and ignore the other. That is not ideal in several ways: a) later updates to the CRL are ignored, b) it fixates on an expired CRL when a "better" one is available, c) the order of CRLs in our API is unspecified, I think.

You're right, though this kinda feels like an independent issue to me. It already wouldn't be picking the "best" CRL, regardless of expiration enforcement. e.g. It doesn't choose the most recently issued one.

Maybe filtering them based on the `thisUpdate` time and choosing the most recent could be an option?

**Comment by jasperpatterson** [comment]
Just wanted to clarify what changes you all would be looking for on this PR in order to move it forward?

**Comment by cpu** [comment]
> Just wanted to clarify what changes you all would be looking for on this PR in order to move it forward?

@jasperpatterson Apologies for the delay responding, I had missed the GitHub notification for this comment.

From my perspective I think this mainly needs the integration test coverage that @ctz mentioned in [his earlier comment](https://github.com/rustls/webpki/pull/227#pullrequestreview-1868791936). 

Beyond that, it would be helpful to see a draft PR on the Rustls repo that updates the [ServerCertVerifierBuilder](https://docs.rs/rustls/latest/rustls/client/struct.ServerCertVerifierBuilder.html) and [ClientCertVerifierBuilder](ClientCertVerifierBuilder) to use the new option implemented here. It would be nice to see this working end-to-end before we lock in the `webpki` side.

**Comment by cpu** [comment]
@jasperpatterson Is this ready for review after the latest push or should we hold off for a draft PR in Rustls, or further adjustments here?

**Comment by jasperpatterson** [comment]
@cpu This is now ready for review!

I added in the integration tests as requested. That caused quite a large diff unfortunately, as I needed to force regenerate all the test CRLs and signatures, otherwise my new tests would fail with `InvalidCrlSignatureForPublicKey`.

I've also opened a draft PR in rustls: https://github.com/rustls/rustls/pull/1922

**Comment by cpu** [comment]
> @cpu This is now ready for review!

Awesome, thanks for iterating on this. I will review ASAP.

> That caused quite a large diff unfortunately, as I needed to force regenerate all the test CRLs and signatures, otherwise my new tests would fail with InvalidCrlSignatureForPublicKey.

NP, that's expected in this case.

**Review by cpu — APPROVED** [review]
Looks good to me. Thank you for following up with the extra test coverage.

I had some nits/minor comments (and I think the `tests/signatures/` changes need to be backed out) but the overall implementation seems correct.

File: tests/signatures/ed25519_key_and_ed25519_detects_bad_signature.sig.bin
Comment (jasperpatterson) [reply tests/signatures/ed25519_key_and_ed25519_detects_bad_signature.sig.bin:1]: I can't get my new tests to pass without adding the `--force` argument, which causes everything to re-generate. I get `InvalidCrlSignatureForPublicKey` otherwise.

File: tests/generate.py
Comment (cpu) [reply tests/generate.py]: :facepalm: Sorry, didn't realize this was a pre-existing typo. Thanks for tidying it up elsewhere as well.

**Review by cpu — APPROVED** [review]

File: src/verify_cert.rs
Comment (djc) [inline src/verify_cert.rs:142]: Nit: would prefer to order `time` between `path` and `trust_anchor` instead of last, fitting the most-to-least-specific order of arguments.

**Review by djc — APPROVED** [review]
LGTM (I did not review the integration tests).

**Comment by cpu** [comment]
@ctz I think this is ready for you to give a review pass when you have a chance.

**Review by ctz — APPROVED** [review]
Thanks!

**Comment by cpu** [comment]
@jasperpatterson Thanks for your patience with this one :bow: I'm preparing a release with this changeset in https://github.com/rustls/webpki/pull/256

**Pull request merged**
